### PR TITLE
Bugfix: accumulation and suggestion for learning rate finder

### PR DESCRIPTION
**Pull request by SkafteNicki**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## What does this PR do?
Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/1767, by implementing a `skip_begin` parameter (default 10) for the `lr_finder.suggestion` method

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/1726, by only logging learning rate and loss when optimizer is called (as it should be). This also removes `num_accumulation_steps` since this is a duplicate of the `accumulate_grad_batches` trainer arg.

Added tests for both.
 

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/lr_finder.py [hunk 1]
```diff
@@ -320,8 +337,12 @@ class _LRCallback(Callback):
     """ Special callback used by the learning rate finder. This callbacks log
     the learning rate before each batch and log the corresponding loss after
     each batch. """
-    def __init__(self, num_training: int, progress_bar_refresh_rate: bool = False, beta: float = 0.98):
+    def __init__(self, num_training: int,
+                 early_stop_threshold: float = 4.0,
+                 progress_bar_refresh_rate: bool = False,
+                 beta: float = 0.98):
```
Comment (Borda) [inline pytorch_lightning/trainer/lr_finder.py:343]: it would be nice to describe arguments, at least the `beta`

File: pytorch_lightning/trainer/lr_finder.py [hunk 2]
```diff
@@ -332,34 +353,37 @@ def __init__(self, num_training: int, progress_bar_refresh_rate: bool = False, b
...
+                self.progress_bar = tqdm(desc='Finding best initial lr', total=self.num_training)
 
-        self.lrs.append(trainer.lr_schedulers[0]['scheduler'].lr[0])
+            self.lrs.append(trainer.lr_schedulers[0]['scheduler'].lr[0])
 
     def on_batch_end(self, trainer, pl_module):
         """ Called when the training batch ends, logs the calculated loss """
-        if self.progress_bar:
-            self.progress_bar.update()
+        if (trainer.batch_idx + 1) % trainer.accumulate_grad_batches == 0:
```
Comment (Borda) [inline pytorch_lightning/trainer/lr_finder.py:364]: if all is in condition rather define reverse negative condition like:
```
if (trainer.batch_idx + 1) % trainer.accumulate_grad_batches != 0:
    returm
```

**Comment by williamFalcon** [comment]
@SkafteNicki want to sneak this last minute into 0.7.6? we're releasing the rc2 in the next hour

**Comment by williamFalcon** [comment]
need to get these tests to pass

**Comment by Borda** [comment]
> need to get these tests to pass

this is fixed In master now, just rebase it...

File: pytorch_lightning/trainer/lr_finder.py [hunk 3]
```diff
@@ -104,6 +111,12 @@ def lr_find(self,
             trainer.fit(model)
 
         """
+        if num_accumulation_steps is not None:
+            rank_zero_warn("Argument `num_accumulation_steps` has been deprepecated"
+                           " since v0.8.0 and will be removed in v1.0.0. Please"
```
Comment (Borda) [inline pytorch_lightning/trainer/lr_finder.py:116]:
```suggestion
                           " since v0.7.6 and will be removed in 0.9. Please"
```

File: pytorch_lightning/trainer/lr_finder.py [hunk 4]
```diff
@@ -298,30 +309,57 @@ def plot(self, suggest: bool = False, show: bool = False):
...
+        early_stop_threshold: threshold for stopping the search. If the
+                loss at any point is larger than early_stop_threshold*best_loss
+                then the search is stopped. To disable, set to None.
+
+        progress_bar_refresh_rate: rate to refresh the progress bar for
+            the learning rate finder
+
+        beta: smoothing value, the loss being logged is a running average of
+            loss values logged until now. beta controls the forget rate i.e.
+            if beta=0 all past information is ignored.
```
Comment (Borda) [inline pytorch_lightning/trainer/lr_finder.py:354]:
```suggestion

    Args:
        num_training: number of iterations done by the learning rate finder
        early_stop_threshold: threshold for stopping the search. If the
                loss at any point is larger than `early_stop_threshold*best_loss`
                then the search is stopped. To disable, set to `None`.
        progress_bar_refresh_rate: rate to refresh the progress bar for
            the learning rate finder
        beta: smoothing value, the loss being logged is a running average of
            loss values logged until now. `beta` controls the forget rate i.e.
            if `beta=0` all past information is ignored.
```

File: pytorch_lightning/trainer/lr_finder.py [hunk 5]
```diff
@@ -298,30 +309,54 @@ def plot(self, suggest: bool = False, show: bool = False):
...
     the learning rate before each batch and log the corresponding loss after
-    each batch. """
-    def __init__(self, num_training: int, progress_bar_refresh_rate: bool = False, beta: float = 0.98):
+    each batch.
+
+    Args:
+        num_training: number of iterations done by the learning rate finder
+        early_stop_threshold: threshold for stopping the search. If the
+                loss at any point is larger than `early_stop_threshold*best_loss`
+                then the search is stopped. To disable, set to `None`.
```
Comment (Borda) [inline pytorch_lightning/trainer/lr_finder.py:346]:
```suggestion
        early_stop_threshold: threshold for stopping the search. If the
            loss at any point is larger than `early_stop_threshold*best_loss`
            then the search is stopped. To disable, set to `None`.
```

File: pytorch_lightning/trainer/lr_finder.py [hunk 6]
```diff
@@ -298,30 +309,54 @@ def plot(self, suggest: bool = False, show: bool = False):
...
-            return self.results["lr"][min_grad]
+            loss = self.results["loss"][skip_begin:-skip_end]
+            min_grad = (np.gradient(np.array(loss))).argmin()
+            self._optimal_idx = min_grad + skip_begin
+            return self.results["lr"][self._optimal_idx]
         except Exception:
-            log.warning('Failed to compute suggesting for `lr`.'
-                        ' There might not be enough points.')
+            log.exception('Failed to compute suggesting for `lr`.'
+                          ' There might not be enough points.')
```
Comment (Borda) [inline pytorch_lightning/trainer/lr_finder.py:333]:
```suggestion
            log.exception('Failed to compute suggesting for `lr`. There might not be enough points.')
```

File: pytorch_lightning/trainer/lr_finder.py [hunk 7]
```diff
@@ -298,30 +309,54 @@ def plot(self, suggest: bool = False, show: bool = False):
...
+    Args:
+        num_training: number of iterations done by the learning rate finder
+        early_stop_threshold: threshold for stopping the search. If the
+            loss at any point is larger than `early_stop_threshold*best_loss`
+            then the search is stopped. To disable, set to `None`.
+        progress_bar_refresh_rate: rate to refresh the progress bar for
+            the learning rate finder
+        beta: smoothing value, the loss being logged is a running average of
+            loss values logged until now. `beta` controls the forget rate i.e.
+            if `beta=0` all past information is ignored.
```
Comment (Borda) [inline pytorch_lightning/trainer/lr_finder.py:351]:
```suggestion
        num_training: number of iterations done by the learning rate finder
        early_stop_threshold: threshold for stopping the search. If the
            loss at any point is larger than ``early_stop_threshold*best_loss``
            then the search is stopped. To disable, set to ``None``.
        progress_bar_refresh_rate: rate to refresh the progress bar for
            the learning rate finder
        beta: smoothing value, the loss being logged is a running average of
            loss values logged until now. ``beta`` controls the forget rate i.e.
            if ``beta=0`` all past information is ignored.
```

File: pytorch_lightning/trainer/lr_finder.py [hunk 8]
```diff
@@ -81,7 +83,12 @@ def lr_find(self,
                 after each batch. If set to 'exponential', will increase learning
                 rate exponentially.
 
-            num_accumulation_steps: number of batches to calculate loss over.
+            early_stop_threshold: threshold for stopping the search. If the
+                loss at any point is larger than early_stop_threshold*best_loss
+                then the search is stopped. To disable, set to None.
+
+            num_accumulation_steps: deprepecated, number of batches to calculate loss over.
+                Set trainer argument `accumulate_grad_batches` instead.
```
Comment (Borda) [inline pytorch_lightning/trainer/lr_finder.py:91]:
```suggestion
                Set trainer argument ``accumulate_grad_batches`` instead.
```

File: pytorch_lightning/trainer/lr_finder.py [hunk 9]
```diff
@@ -298,30 +309,54 @@ def plot(self, suggest: bool = False, show: bool = False):
...
         as the point with the steepest negative gradient.
 
         Returns:
             lr: suggested initial learning rate to use
 
+            skip_begin: how many samples to skip in the beginning.
+                Prevent too naive estimates
+
+            skip_end: how many samples to skip in the end.
+                Prevent too optimistic estimates
```
Comment (Borda) [inline pytorch_lightning/trainer/lr_finder.py:323]:
```suggestion
            skip_begin: how many samples to skip in the beginning.
                Prevent too naive estimates
            skip_end: how many samples to skip in the end.
                Prevent too optimistic estimates
```

File: pytorch_lightning/trainer/lr_finder.py [hunk 10]
```diff
@@ -298,30 +309,51 @@ def plot(self, suggest: bool = False, show: bool = False):
...
+    def suggestion(self, skip_begin: int = 10, skip_end: int = 1):
         """ This will propose a suggestion for choice of initial learning rate
         as the point with the steepest negative gradient.
 
         Returns:
             lr: suggested initial learning rate to use
+            skip_begin: how many samples to skip in the beginning.
+                Prevent too naive estimates
+            skip_end: how many samples to skip in the end.
+                Prevent too optimistic estimates
```
Comment (Borda) [inline pytorch_lightning/trainer/lr_finder.py:321]:
```suggestion
            skip_begin: how many samples to skip in the beginning. Prevent too naive estimates
            skip_end: how many samples to skip in the end. Prevent too optimistic estimates
```
